### PR TITLE
Add Map support to (Android) code generation

### LINKED ISSUE #1754: Code generation for plain maps isn't working on Android
The Java code, generated for APIs isn't complete if there are properties/parameters with `object` defined the type. Let's say I define an event like this:

```json
...
  "paths": {
    "event/demo": {
      "event": {
        "tags": ["Demo"],
        "operationId": "Demo",
        "description": "Demo",
        "parameters": [
          {
            "name": "param1",
            "in": "body",
            "description": "This is param1",
            "schema": {
              "type": "string"
            }
          },
          {
            "name": "param2",
            "in": "body",
            "description": "This is param2",
            "schema": {
              "type": "object"
            }
          }
        ]
      }
    }
  },
  ...
```
The event has 2 parameters, one is a `string` and the other is an `object`. After running `ern regen-api`, this will result in the following iOS (intializer) code:

```swift
    public let param1: String?
    public let param2: AnyObject?
    
    public required init(dictionary: [AnyHashable: Any]) {
        if
        {
        } else {
            assertionFailure("\(DemoData.tag) missing one or more required properties[]")
        }

        param1 = dictionary["param1"] as? String
        param2 = dictionary["param2"] as? AnyObject

        super.init(dictionary: dictionary)
    }
```
As you can see, both params are properly decoded (although I'm not sure `AnyObject` is the most optimal type, would probably go for `[AnyHashable: Any]`).

The problem, however, is in the generated Android code:
```java
    private String param1;
    private Object param2;
    
    public DemoData(@NonNull Bundle bundle) {
        this.param1 = bundle.getString("param1");
    }

```
Only the first parameter of type `string` is properly initialized but the second property of type `object` is never set.

### PR DESCRIPTION
**Pull request by LcTwisk**
This fixes #1754 in which `object` type params/properties, defined in API schema's,  don't result in any Java code.

The reason it's being skipped is that all different primitive types are handled separately in the template, but plain maps seem to be missing. On the Swift side, this doesn't seem to be a problem because all primitive types are rendered in the same way in the template.

In order to render the property initialization for `object` types in Java, I've added a new property to the rendering view, called `isObject`.  Additionally, I've added a matching condition to the template. 

Aside from that, I took the freedom to map a JS `object` to a `Bundle` type in Java. Previously, this was represented as an `Object` in Java, which might be a bit too generic, the bridge also uses `Bundle` as map type internally.
Can't see why this wouldn't work, please let me know if I'm missing something.

The new Java output for the schema defined in #1754 is now:
```java
    private String param1;
    private Bundle param2;
    
    public DemoData(@NonNull Bundle bundle) {
        this.param1 = bundle.getString("param1");
        this.param2 = bundle.containsKey("param2") ? bundle.getBundle("param2") : null;
    }
```
Now both properties are being initialized.

### CODE REVIEW

**Review by friederbluemle — COMMENTED** [review]
Thanks @LcTwisk - This looks like like a nice improvement! :+1:

I left one comment about Bundle - Not sure if it can be done differently, please check when you have some time.

File: ern-api-gen/src/DefaultCodegen.ts [hunk 1]
```diff
@@ -210,7 +211,7 @@ export default class DefaultCodegen {
     ['short', 'Short'],
     ['char', 'String'],
     ['double', 'Double'],
-    ['object', 'Object'],
```
[the remark this answers is not in the document]
Comment (LcTwisk) [reply ern-api-gen/src/DefaultCodegen.ts:213]: Ahh, that would make sense for this change indeed. Regarding the order change of adding `isObject` to the view, I guess that's fine to keep right as it's not Android specific?

[the remark this answers is not in the document]
Comment (LcTwisk) [reply ern-api-gen/src/DefaultCodegen.ts]: @friederbluemle This is taken care of! I'm now overriding the `object` entry in `AndroidClientCodegen.ts`: https://github.com/electrode-io/electrode-native/blob/800596e82da12e6f11a956a721c4d4d58567ef99/ern-api-gen/src/languages/AndroidClientCodegen.ts#L129

[the remark this answers is not in the document]
Comment (friederbluemle) [reply ern-api-gen/src/DefaultCodegen.ts]: Awesome thanks :+1:

**Review by friederbluemle — APPROVED** [review]
LGTM thanks @LcTwisk 👍 Can you rebase the branch please?

@belemaire - Can you also take a quick look?

**Comment by belemaire** [comment]
Yes I will have a look today

**Review by belemaire — APPROVED** [review]
Thanks much @LcTwisk LGTM :+1:

ERN API generator module is not a super easy one to maintain, as you may have noticed, so such improvements are greatly appreciated.

One thing that would be a nice addition to this PR would be to introduce a new system test so that we can see the corresponding generated code _(and also avoid the risk of breaking this functionality in the future)_. If you don't feel like doing it, that's ok we will take care of it via a follow up PR.

Unfortunately this is not super well documented but is relatively straightforward.
Just add the following JSON _(adapted from the one you provided in https://github.com/electrode-io/electrode-native/issues/1754)_ before [this line in complexapi-schema.json]( https://github.com/electrode-io/electrode-native/blob/master/system-tests/fixtures/api/complexapi-schema.json#L11)

```json
"event/testObjectParam": {
  "event": {
    "tags": ["TestEventObjectParam"],
    "operationId": "TestEventObjectParam",
    "description": "TestEventObjectParam",
    "parameters": [
      {
        "name": "param1",
        "in": "body",
        "description": "This is param1",
        "schema": {
          "type": "string"
        }
      },
      {
        "name": "param2",
        "in": "body",
        "description": "This is param2",
        "schema": {
          "type": "object"
        }
      }
    ]
  }
}
```

And then from `electrode-native` repository root run `yarn regen-fixtures` and only select `complex-api` from the list. This will regenerate this API test fixture that we use during system tests. Then just rebase & commit back.

Thanks @friederbluemle for initial review

**Comment by LcTwisk** [comment]
> Thanks much @LcTwisk LGTM 👍
> 
> ERN API generator module is not a super easy one to maintain, as you may have noticed, so such improvements are greatly appreciated.
> 
> One thing that would be a nice addition to this PR would be to introduce a new system test so that we can see the corresponding generated code _(and also avoid the risk of breaking this functionality in the future)_. If you don't feel like doing it, that's ok we will take care of it via a follow up PR.
> 
> Unfortunately this is not super well documented but is relatively straightforward.
> Just add the following JSON _(adapted from the one you provided in #1754)_ before [this line in complexapi-schema.json](https://github.com/electrode-io/electrode-native/blob/master/system-tests/fixtures/api/complexapi-schema.json#L11)
> 
> ```json
> "event/testObjectParam": {
>   "event": {
>     "tags": ["TestEventObjectParam"],
>     "operationId": "TestEventObjectParam",
>     "description": "TestEventObjectParam",
>     "parameters": [
>       {
>         "name": "param1",
>         "in": "body",
>         "description": "This is param1",
>         "schema": {
>           "type": "string"
>         }
>       },
>       {
>         "name": "param2",
>         "in": "body",
>         "description": "This is param2",
>         "schema": {
>           "type": "object"
>         }
>       }
>     ]
>   }
> }
> ```
> 
> And then from `electrode-native` repository root run `yarn regen-fixtures` and only select `complex-api` from the list. This will regenerate this API test fixture that we use during system tests. Then just rebase & commit back.
> 
> Thanks @friederbluemle for initial review

Thanks for the review @belemaire will make sure to add a system test somewhere in the coming days 👍

**Comment by belemaire** [comment]
@LcTwisk
We're releasing 0.46.0 today, and felt that this fix should be included in this this release.
I have rebased your branch and updated API schema of system test to add a use case for map as discussed.
Thanks again for the contribution :+1:

**Pull request merged**